File: apps/runtime/src/media/ffmpeg_compose.cpp
#include "omega/runtime/media/ffmpeg_compose.hpp"

#include "omega/runtime/media/ffmpeg_text_overlay.hpp"

#include <algorithm>
#include <cstdio>
#include <fstream>
#include <sstream>

#ifdef _WIN32
#ifndef NOMINMAX
#define NOMINMAX
#endif
#include <windows.h>
#else
#include <unistd.h>
#endif

namespace fs = std::filesystem;
using json = nlohmann::json;

namespace omega::runtime::ffmpeg_media {

namespace {

int run_cmd(const std::string& cmd) {
  return std::system(cmd.c_str());
}

std::string quote(const fs::path& p) {
  const std::string s = p.string();
  if (s.find(' ') == std::string::npos && s.find('"') == std::string::npos) return s;
  std::string out = "\"";
  for (char c : s) {
    if (c == '"') out += "\\\"";
    else out += c;
  }
  out += "\"";
  return out;
}

std::pair<int, int> dims_from_aspect(const std::string& aspect) {
  if (aspect == "9:16" || aspect == "vertical") return {720, 1280};
  return {1280, 720};
}

std::vector<json> sorted_scenes(const json& script) {
  std::vector<json> scenes;
  if (!script.contains("scenes") || !script["scenes"].is_array()) return scenes;
  for (const auto& sc : script["scenes"]) scenes.push_back(sc);
  std::sort(scenes.begin(), scenes.end(), [](const json& a, const json& b) {
    return a.value("scene_number", 0) < b.value("scene_number", 0);
  });
  return scenes;
}

std::string concat_line(const fs::path& p) {
  std::string text = fs::absolute(p).lexically_normal().string();
  for (char& c : text) {
    if (c == '\\') c = '/';
  }
  std::string out = "file '";
  for (char c : text) {
    if (c == '\'') out += "'\\''";
    else out += c;
  }
  out += "'";
  return out;
}

}  // namespace

std::string find_ffmpeg() {
#ifdef _WIN32
  const char* names[] = {"ffmpeg.exe", "ffmpeg"};
#else
  const char* names[] = {"ffmpeg"};
#endif
  if (const char* path = std::getenv("PATH")) {
    std::string paths = path;
    size_t start = 0;
    while (start < paths.size()) {
#ifdef _WIN32
      const char sep = ';';
#else
      const char sep = ':';
#endif
      const size_t pos = paths.find(sep, start);
      const size_t e = pos == std::string::npos ? paths.size() : pos;
      const std::string dir = paths.substr(start, e - start);
      for (const char* name : names) {
        const fs::path candidate = fs::path(dir) / name;
        if (fs::exists(candidate)) return candidate.string();
      }
      if (pos == std::string::npos) break;
      start = pos + 1;
    }
  }
  return "ffmpeg";
}

std::string find_ffprobe() {
#ifdef _WIN32
  const char* names[] = {"ffprobe.exe", "ffprobe"};
#else
  const char* names[] = {"ffprobe"};
#endif
  if (const char* path = std::getenv("PATH")) {
    std::string paths = path;
    size_t start = 0;
    while (start < paths.size()) {
#ifdef _WIN32
      const char sep = ';';
#else
      const char sep = ':';
#endif
      const size_t pos = paths.find(sep, start);
      const size_t e = pos == std::string::npos ? paths.size() : pos;
      const std::string dir = paths.substr(start, e - start);
      for (const char* name : names) {
        const fs::path candidate = fs::path(dir) / name;
        if (fs::exists(candidate)) return candidate.string();
      }
      if (pos == std::string::npos) break;
      start = pos + 1;
    }
  }
  return "ffprobe";
}

bool write_silent_wav(const fs::path& wav_path, const double duration_seconds) {
  const std::string ffmpeg = find_ffmpeg();
  if (ffmpeg.empty()) return false;
  const double dur = std::max(0.5, duration_seconds);
  fs::create_directories(wav_path.parent_path());
  std::ostringstream cmd;
  cmd << quote(fs::path(ffmpeg)) << " -y -f lavfi -i anullsrc=r=24000:cl=mono -t " << dur << " "
      << quote(wav_path) << " 2>nul";
  return run_cmd(cmd.str()) == 0 && fs::exists(wav_path);
}

int pcm16_wav_peak_abs(const fs::path& wav_path) {
  std::ifstream in(wav_path, std::ios::binary);
  if (!in) return -1;
  char header[44];
  if (!in.read(header, 44)) return -1;
  if (std::string(header, 4) != "RIFF" || std::string(header + 8, 4) != "WAVE") return -1;
  int peak = 0;
  int16_t sample = 0;
  while (in.read(reinterpret_cast<char*>(&sample), sizeof(sample))) {
    const int abs_s = sample < 0 ? -static_cast<int>(sample) : static_cast<int>(sample);
    if (abs_s > peak) peak = abs_s;
  }
  return peak;
}

int count_nearly_silent_scene_wavs(const fs::path& audio_dir,
                                   const std::vector<json>& scenes) {
  if (scenes.empty()) return 0;
  int silent = 0;
  for (size_t i = 0; i < scenes.size(); ++i) {
    const json& sc = scenes[i];
    const int sn = sc.value("scene_number", static_cast<int>(i + 1));
    char name_buf[32];
    std::snprintf(name_buf, sizeof(name_buf), "scene_%02d", sn);
    const fs::path wav = audio_dir / (std::string(name_buf) + ".wav");
    const int peak = pcm16_wav_peak_abs(wav);
    if (peak >= 0 && peak <= 8) ++silent;
  }
  return silent;
}

double probe_duration_seconds(const fs::path& media_path) {
  const std::string ffprobe = find_ffprobe();
  std::ostringstream cmd;
  cmd << quote(fs::path(ffprobe)) << " -v error -show_entries format=duration "
      << "-of default=noprint_wrappers=1:nokey=1 " << quote(media_path) << " 2>nul";
  FILE* pipe = nullptr;
#ifdef _WIN32
  pipe = _popen(cmd.str().c_str(), "r");
#else
  pipe = popen(cmd.str().c_str(), "r");
#endif
  if (!pipe) return 1.0;
  char buf[128] = {};
  if (!fgets(buf, sizeof(buf), pipe)) {
#ifdef _WIN32
    _pclose(pipe);
#else
    pclose(pipe);
#endif
    return 1.0;
  }
#ifdef _WIN32
  _pclose(pipe);
#else
  pclose(pipe);
#endif
  try {
    return std::max(0.1, std::stod(buf));
  } catch (...) {
    return 1.0;
  }
}

fs::path assemble_final_mp4(const fs::path& job_root, const json& script_content,
                            const json& brief_json, std::string& error,
                            const bool include_subtitles) {
  const std::string ffmpeg = find_ffmpeg();
  if (ffmpeg.empty()) {
    error = "ffmpeg not found on PATH";
    return {};
  }

  const auto scenes = sorted_scenes(script_content);
  if (scenes.empty()) {
    error = "no scenes in script";
    return {};
  }

  const std::string aspect = brief_json.value("aspect_ratio", "16:9");
  const auto [tw, th] = dims_from_aspect(aspect);
  const fs::path images_dir = job_root / "images";
  const fs::path audio_dir = job_root / "audio";
  const fs::path segments_dir = job_root / "segments";
  fs::create_directories(segments_dir);

  const fs::path caption_font = segments_dir / "caption_font.ttf";
  std::string fontfile_arg;
  if (const auto sys_font = resolve_system_font_path()) {
    if (!fs::exists(caption_font)) {
      std::error_code ec;
      fs::copy_file(*sys_font, caption_font, fs::copy_options::overwrite_existing, ec);
    }
    if (fs::exists(caption_font)) fontfile_arg = "caption_font.ttf";
  }

  std::vector<fs::path> segment_paths;
  for (size_t i = 0; i < scenes.size(); ++i) {
    const json& sc = scenes[i];
    const int sn = sc.value("scene_number", static_cast<int>(i + 1));
    char name_buf[32];
    std::snprintf(name_buf, sizeof(name_buf), "scene_%02d", sn);
    const fs::path img = images_dir / (std::string(name_buf) + ".png");
    const fs::path wav = audio_dir / (std::string(name_buf) + ".wav");
    if (!fs::exists(img)) {
      error = "missing image for scene " + std::to_string(sn);
      return {};
    }
    if (!fs::exists(wav)) {
      error = "missing audio for scene " + std::to_string(sn);
      return {};
    }

    const fs::path seg = segments_dir / (std::string(name_buf) + ".mp4");
    const double audio_dur = probe_duration_seconds(wav);
    std::string base_vf = "scale=" + std::to_string(tw) + ":" + std::to_string(th) +
                          ":force_original_aspect_ratio=decrease,pad=" + std::to_string(tw) + ":" +
                          std::to_string(th) +
                          ":(ow-iw)/2:(oh-ih)/2:color=0x1a1a1a,format=yuv420p";
    if (include_subtitles) {
      const std::string caption = overlay_caption_for_scene(sc);
      const std::string caption_file =
          fontfile_arg.empty() ? std::string{} : write_caption_textfile(segments_dir, sn, caption);
      base_vf =
          video_filter_with_caption(base_vf, caption, tw, th, fontfile_arg, caption_file);
    }

    std::ostringstream cmd;
#ifdef _WIN32
    cmd << "cd /d " << quote(segments_dir) << " && ";
#else
    cmd << "cd " << quote(segments_dir) << " && ";
#endif
    cmd << quote(fs::path(ffmpeg)) << " -y -loop 1 -framerate 25 -i " << quote(img) << " -i "
        << quote(wav) << " -map 0:v:0 -map 1:a:0 -vf \"" << base_vf << "\" -c:v libx264 -preset "
        << "medium -crf 23 -c:a aac -b:a 192k -ar 48000 -ac 1 -t " << audio_dur << " -movflags "
        << "+faststart " << quote(seg) << " 2>&1";
    if (run_cmd(cmd.str()) != 0) {
      error = "ffmpeg scene encode failed for scene " + std::to_string(sn);
      return {};
    }
    segment_paths.push_back(seg);
  }

  const fs::path final_mp4 = job_root / "final.mp4";
  if (segment_paths.size() == 1) {
    std::error_code ec;
    fs::copy_file(segment_paths[0], final_mp4, fs::copy_options::overwrite_existing, ec);
    if (ec) {
      error = "failed to copy single segment: " + ec.message();
      return {};
    }
    return final_mp4;
  }

  const fs::path concat_list = segments_dir / "concat.txt";
  {
    std::ofstream out(concat_list);
    for (const auto& p : segment_paths) out << concat_line(p) << "\n";
  }

  std::ostringstream cmd;
  cmd << quote(fs::path(ffmpeg)) << " -y -f concat -safe 0 -i " << quote(concat_list)
      << " -c:v libx264 -preset medium -crf 23 -c:a aac -b:a 192k -ar 48000 -ac 2 -movflags "
      << "+faststart " << quote(final_mp4) << " 2>&1";
  if (run_cmd(cmd.str()) != 0) {
    error = "ffmpeg concat failed";
    return {};
  }
  return final_mp4;
}

}  // namespace omega::runtime::ffmpeg_media

File: apps/desktop/content-studio/backend/app/services/ffmpeg_compose.py
"""Assemble scene PNG + WAV segments into one MP4 via ffmpeg."""

from __future__ import annotations

import shutil
import subprocess
from pathlib import Path
from typing import Any

from sqlalchemy.orm import Session

from app.config import settings
from app.models import JobLog
from app.services.ffmpeg_fonts import resolve_system_font_path
from app.services.ffmpeg_text_overlay import (
    overlay_caption_for_scene,
    video_filter_with_caption,
    write_caption_textfile,
)


def _job_root(project_id: str, job_id: str) -> Path:
    return Path(settings.storage_path).expanduser().resolve() / project_id / job_id


def _dims_from_aspect(aspect: str) -> tuple[int, int]:
    a = (aspect or "16:9").strip()
    if a in ("9:16", "vertical"):
        return 720, 1280
    return 1280, 720


from app.services.script_scenes import sorted_script_scenes as _sorted_script_scenes


def _ffmpeg_concat_line(path: Path) -> str:
    """Absolute path safe for ffmpeg concat demuxer (forward slashes, escaped quotes)."""
    text = path.resolve().as_posix().replace("'", "'\\''")
    return f"file '{text}'"


def ffprobe_duration_seconds(media_path: Path) -> float:
    exe = shutil.which("ffprobe")
    if not exe:
        raise RuntimeError("ffprobe not found on PATH (install FFmpeg).")
    r = subprocess.run(
        [
            exe,
            "-v",
            "error",
            "-show_entries",
            "format=duration",
            "-of",
            "default=noprint_wrappers=1:nokey=1",
            str(media_path),
        ],
        capture_output=True,
        text=True,
        check=True,
    )
    return float(r.stdout.strip())


def ffprobe_has_audio_stream(media_path: Path) -> bool:
    exe = shutil.which("ffprobe")
    if not exe:
        return True
    r = subprocess.run(
        [
            exe,
            "-v",
            "error",
            "-select_streams",
            "a",
            "-show_entries",
            "stream=codec_type",
            "-of",
            "csv=p=0",
            str(media_path),
        ],
        capture_output=True,
        text=True,
    )
    return bool((r.stdout or "").strip())


def assemble_final_mp4(
    db: Session,
    *,
    job_id: str,
    project_id: str,
    script_content: dict[str, Any],
    brief_json: dict[str, Any],
) -> Path:
    """Concatenate per-scene MP4s (still image + narration) into ``final.mp4``."""
    ffmpeg = shutil.which("ffmpeg")
    if not ffmpeg:
        raise RuntimeError("ffmpeg not found on PATH (install FFmpeg and restart).")

    scenes = _sorted_script_scenes(script_content)
    if not scenes:
        raise RuntimeError("No scenes to render.")

    root = _job_root(project_id, job_id)
    audio_dir = root / "audio"
    images_dir = root / "images"
    segments_dir = root / "segments"
    segments_dir.mkdir(parents=True, exist_ok=True)

    aspect = str(brief_json.get("aspect_ratio") or "16:9")
    tw, th = _dims_from_aspect(aspect)
    font_path = resolve_system_font_path()
    caption_font = segments_dir / "caption_font.ttf"
    if font_path and not caption_font.is_file():
        shutil.copy2(font_path, caption_font)
    fontfile = "caption_font.ttf" if caption_font.is_file() else None
    db.add(
        JobLog(
            job_id=job_id,
            level="info",
            message=(
                f"ffmpeg: assembling {len(scenes)} scene(s) "
                f"({tw}x{th})"
                + (f", caption font: {caption_font.name}" if fontfile else ", captions skipped (no system font)")
            ),
        )
    )
    db.commit()

    segment_paths: list[Path] = []
    for i, sc in enumerate(scenes):
        sn = int(sc.get("scene_number") or i + 1)
        img = images_dir / f"scene_{sn:02d}.png"
        wav = audio_dir / f"scene_{sn:02d}.wav"
        if not img.is_file():
            raise RuntimeError(f"Missing image for scene {sn}: {img}")
        if not wav.is_file():
            raise RuntimeError(f"Missing audio for scene {sn}: {wav}")
        seg = segments_dir / f"scene_{sn:02d}.mp4"
        audio_dur = max(0.1, float(ffprobe_duration_seconds(wav)))
        # Letterbox pad uses dark gray (not black) when aspect differs from still.
        base_vf = (
            f"scale={tw}:{th}:force_original_aspect_ratio=decrease,"
            f"pad={tw}:{th}:(ow-iw)/2:(oh-ih)/2:color=0x1a1a1a,format=yuv420p"
        )
        caption = overlay_caption_for_scene(sc)
        caption_fontsize = max(28, min(56, th // 18))
        caption_file = (
            write_caption_textfile(segments_dir, sn, caption, width=tw, fontsize=caption_fontsize)
            if fontfile
            else None
        )
        vf = video_filter_with_caption(
            base_vf,
            caption,
            width=tw,
            height=th,
            fontfile=fontfile,
            textfile=caption_file,
        )
        cmd = [
            ffmpeg,
            "-y",
            "-loop",
            "1",
            "-framerate",
            "25",
            "-i",
            str(img.resolve()),
            "-i",
            str(wav.resolve()),
            "-map",
            "0:v:0",
            "-map",
            "1:a:0",
            "-vf",
            vf,
            "-c:v",
            "libx264",
            "-preset",
            "medium",
            "-crf",
            "23",
            "-c:a",
            "aac",
            "-b:a",
            "192k",
            "-ar",
            "48000",
            "-ac",
            "1",
            "-t",
            f"{audio_dur:.3f}",
            "-movflags",
            "+faststart",
            str(seg.resolve()),
        ]
        db.add(JobLog(job_id=job_id, level="info", message=f"ffmpeg: encoding scene {sn} segment…"))
        db.commit()
        r = subprocess.run(cmd, capture_output=True, text=True, cwd=str(segments_dir))
        if r.returncode != 0:
            err = (r.stderr or r.stdout or "ffmpeg scene encode failed").strip()
            tail = err[-2500:] if len(err) > 2500 else err
            db.add(
                JobLog(
                    job_id=job_id,
                    level="error",
                    message=f"ffmpeg scene {sn} failed: {tail}",
                )
            )
            db.commit()
            raise RuntimeError(
                f"ffmpeg scene {sn} encode failed"
                + (f" (Fontconfig/drawtext — rebuild Omega for caption font fix). {tail[-400:]}" if "Fontconfig" in err or "drawtext" in err else f". {tail[-400:]}")
            )
        if not ffprobe_has_audio_stream(seg):
            db.add(
                JobLog(
                    job_id=job_id,
                    level="error",
                    message=(
                        f"ffmpeg: scene {sn} segment has no audio stream — "
                        f"check TTS output at {wav.name} (silent placeholder or bad WAV)."
                    ),
                )
            )
            db.commit()
        segment_paths.append(seg)

    final = root / "final.mp4"
    if len(segment_paths) == 1:
        shutil.copy2(segment_paths[0], final)
    else:
        concat_list = segments_dir / "concat.txt"
        concat_list.write_text(
            "\n".join(_ffmpeg_concat_line(p) for p in segment_paths),
            encoding="utf-8",
        )
        # Re-encode on concat — stream copy often drops AAC audio on Windows players.
        cmd = [
            ffmpeg,
            "-y",
            "-f",
            "concat",
            "-safe",
            "0",
            "-i",
            str(concat_list),
            "-c:v",
            "libx264",
            "-preset",
            "medium",
            "-crf",
            "23",
            "-c:a",
            "aac",
            "-b:a",
            "192k",
            "-ar",
            "48000",
            "-ac",
            "2",
            "-movflags",
            "+faststart",
            str(final),
        ]
        r = subprocess.run(cmd, capture_output=True, text=True)
        if r.returncode != 0:
            err = (r.stderr or r.stdout or "ffmpeg concat failed").strip()
            raise RuntimeError(err[-2000:] if len(err) > 2000 else err)

    if not ffprobe_has_audio_stream(final):
        db.add(
            JobLog(
                job_id=job_id,
                level="error",
                message=(
                    "ffmpeg: final MP4 has no audio track — narration WAVs may be silent placeholders. "
                    "Check job logs for Local TTS warnings and confirm a TTS model is installed in Settings."
                ),
            )
        )
        db.commit()
    else:
        db.add(
            JobLog(
                job_id=job_id,
                level="info",
                message="ffmpeg: final MP4 includes an audio track.",
            )
        )
        db.commit()

    db.add(JobLog(job_id=job_id, level="info", message=f"ffmpeg: final MP4 → {final}"))
    db.commit()
    return final
